libminifi/test/archive-tests/CompressContentTests.cpp

```diff
@@ -91,851 +91,826 @@ class ReadCallback: public org::apache::nifi::minifi::InputStreamCallback {
   int archive_buffer_size_;
 };
 
-TEST_CASE("CompressFileGZip", "[compressfiletest1]") {
-  try {
-    std::ofstream expectfile;
-    expectfile.open(EXPECT_COMPRESS_CONTENT);
+class CompressDecompressionTestController : public TestController{
+protected:
+  static std::string tempDir;
+  static std::string raw_content_path;
+  static std::string compressed_content_path;
+  static TestController global_controller;
+public:
+  class RawContent{
+    std::string content_;
+    RawContent(std::string&& content_): content_(std::move(content_)) {}
+    friend class CompressDecompressionTestController;
+  public:
+    bool operator==(const std::string& actual) const noexcept {
+      return content_ == actual;
+    }
+    bool operator!=(const std::string& actual) const noexcept {
+      return content_ != actual;
+    }
+  };
 
-    std::mt19937 gen(std::random_device { }());
+  std::string rawContentPath() const {
+    return raw_content_path;
+  }
+
+  std::string compressedPath() const {
+    return compressed_content_path;
+  }
+
+  RawContent getRawContent() const {;
+    std::ifstream file;
+    file.open(raw_content_path, std::ios::binary);
+    std::string contents((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+    file.close();
+    return {std::move(contents)};
+  }
+
+  virtual ~CompressDecompressionTestController() = 0;
+};
+
+CompressDecompressionTestController::~CompressDecompressionTestController() {}
```

Comment:
agree with the `default` but why shouldn't it be pure virtual, being pure virtual and having an implementation are orthogonal

Reply:
In my understanding, pure virtual means that the function could not be implemented in the base class, therefore it has to be implemented in the derived class(es). This implies that being pure virtual and having an implementation is contradictory. Pure virtual, i.e. "not being able to implement here" doesn't make sense with destructors, since they have to be implemented in every class of the hierarchy.

Reply:
having a pure virtual method (either declared in the class or inherited) only means that the class cannot be instantiated, you can have an out-of-class definition for such methods and reach them through static dispatch

Reply:
You're right. I did some research and these are indeed the rules. I have never seen this technique before, but I'm fine with it now.

Could you write a code comment describing why you want to make this class abstract without actually enforcing the overriding of any behavior? What is the intention behind this design choice?

Reply:
Thanks for teaching me something new. :)

Reply:
added comment

Reply:
This seems fancy and all, but what is the advantage of having an implementation **if** the implementation is `{}` anyways?

Reply:
you cannot have a class with no destructor defined

Reply:
@hunyadi-dev : https://www.studytonight.com/cpp/virtual-destructors.php#:~:text=Pure%20Virtual%20Destructors%20are%20legal,create%20object%20of%20that%20class.

It must be pure virtual to avoid instantiation, but the dtor must have a body. (As it will get called when you delete a derived type).

Reply:
@arpadboda, @adebreceni Maybe I phrased my comment the wrong way. As far as I understand this is the same as:
```c++
CompressDecompressionTestController() = delete;
virtual ~CompressDecompressionTestController() = default;
```
expect it enforces destructor implementation. What is the advantage of doing this enforcement?

Reply:
this would make it impossible to instantiate not only this class but also any derived classes, an equivalent solution would be to mark the constructor protected, but since we are playing the inheritance game, it is wise to mark the destructor virtual anyway

Reply:
Just posting some references:

> A virtual function declared in a class shall be defined, or declared pure ([class.abstract]) in that class, **or both**; no diagnostic is required ([basic.def.odr]).

http://eel.is/c++draft/class.virtual#12 (emphasis mine)


> The definition of a pure virtual function may be provided (and must be provided if the pure virtual is the destructor): the member functions of the derived class are free to call the abstract base's pure virtual function using qualified function id. This definition must be provided outside of the class body (the syntax of a function declaration doesn't allow both the pure specifier = 0 and a function body).

https://en.cppreference.com/w/cpp/language/abstract_class

Reply:
@adebreceni Ah, now I see why the need for having both, I thought it was possible to exprect abstactness	somewhere else than the destructor, but now I am convinced otherwise.